Public key encryption depends on some pretty deep maths, but it boils down to this

Anyone can encode a message to you using your *public key*, a pair of numbers `r` and `n`.

Only you can decode the message using your *private key*, another number `s`.

It depends on a relationship between the original message `M` and the encoded message `C`

For three carefully chosen numbers `n`, `r` and `s`,

If `M`<sup>`r`</sup>` %n = C` then `C`<sup>`s`</sup>` %n = M`

To encode, raise the message to the power `r` and find the remainder when you divide by `n`

To decode, raise the coded message to the power `s` and find the remainder when you divide by `n`

Choosing `n`, `r` and `s`

* `n` is the product of two primes `p` and `q`
* `r` has no common factors with `m`, the lowest common multiple of `(p-1)` and `(q-1)`
* `rs %m = 1`

### Key exchange

The numbers `n` and `r` are your *public key*. Anyone can use them to enode a message to you.

The numbers `n` and `s` are your *private key*. Only you know them, so only you can decode the message.

### Generating your public key

Pick two prime numbers `p` and `q` and multiply them to get `n=pq`

In [1]:
p = 19
q = 29
n = p*q
n

551

For the public key we need `r` to have no common factors with `(p-1)(q-1)`

So we'll find the lowest common multiple of `(p-1)(q-1)` using the `lcm` function from the `sympy` module, and call it `m`

In [2]:
from sympy import lcm

In [3]:
m = lcm(p-1,q-1)
m

252

We can choose `r=5` because `252` is not divisible by `5`

In [4]:
r = 5

The numbers `n=551` and `r=5` are our *public key*

For our *private key*, we need another number `s` such that `rs%m=1` is one more than a multiple of `m`.

In other words, `rs=km+1`

In [6]:
from sympy import diophantine, symbols
s , k = symbols('s k')

In [ ]:
diophantine(r*s - k*m - 1)

This tells us we could choose any number of the form `252t+101` for our `s`, so let's choose `s=252+101=353`

In [8]:
s = 252 + 101
s

353

Just to check:

In [9]:
r*s%m

1

We have everything we need

* A public key, `n=551` and `r=5`
* A private key, `n=551` and `s=353`

We give everyone our public key, so they generate the *coded* message `C` by doing `M` to the power `r` mod `n`

Suppose someone wanted to send you the message `7`, so `M=7`

In [ ]:
M = 7

They use your public key `n=551` and `r=5`, and do `C = M`<sup>`r`</sup>`%n`

In [ ]:
C = M**r%n
C

When we receive `C` we can `decode` by doing `C` to the power `s` mod `n`

In [12]:
(C**s)%n

7

The message was `7`. Phew.